raft/src/main/java/org/apache/kafka/raft/QuorumState.java

```diff
@@ -125,9 +125,16 @@ public void initialize(OffsetAndEpoch logEndOffsetAndEpoch) throws IOException,
         final EpochState initialState;
         if (!election.voters().isEmpty() && !voters.equals(election.voters())) {
             throw new IllegalStateException("Configured voter set: " + voters
-                + " is different from the voter set read from the state file: " + election.voters() +
-                ". Check if the quorum configuration is up to date, " +
-                "or wipe out the local state file if necessary");
+                + " is different from the voter set read from the state file: " + election.voters()
+                + ". Check if the quorum configuration is up to date, "
+                + "or wipe out the local state file if necessary");
+        } else if (election.hasVoted() && !isVoter()) {
+            String localIdDescription = localId.isPresent() ?
+                localId.getAsInt() + " is not a voter" :
+                "is undefined";
+            throw new IllegalStateException("Initialized quorum state " + election
```

Comment:
So a node could not transit from follower to observer though restarts?

Reply:
I thought I would start out strict and relax later if needed. Worst case, the user would just remove or edit the `quorum-state` file.

Reply:
Fair enough, we could relax later when we experiment out the static quorum changes.